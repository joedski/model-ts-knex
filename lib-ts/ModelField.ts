import { Raw } from "knex";

type ColumnType =
  | ['integer']
  | ['bigInteger']
  /**
   * there's an optional 'medium' | 'big' that could go after,
   * but that's for mysql only, and I'm using postgres.
   */
  | ['text']
  | ['string', number | null]
  | ['float', number | null, number | null]
  | ['decimal', number | null, number | null]
  | ['boolean']
  | ['date']
  | ['datetime', { useTz?: boolean; precision?: number }]
  | ['time', number | null]
  | ['timestamp', { useTz?: boolean; precision?: number }]
  | ['binary']
  | ['enum', Array<string> | null, { useNative?: boolean; enumName?: string }]
  | ['json']
  | ['jsonb']
  | ['uuid']
  ;

export default class ModelField<TSType, InNew extends boolean = true, NotNullable extends boolean = false> {
  // Type Factory Helpers

  public static integer() {
    const field = new ModelField<number>();
    field.$type = ['integer'];
    return field;
  }
  public static bigInteger() {
    const field = new ModelField<number>();
    field.$type = ['bigInteger'];
    return field;
  }
  public static text() {
    const field = new ModelField<string>();
    field.$type = ['text'];
    return field;
  }
  public static string(length: number = 256) {
    const field = new ModelField<string>();
    field.$type = ['string', length];
    return field;
  }
  public static float(precision?: number, scale?: number) {
    const field = new ModelField<number>();
    field.$type = ['float', typeof precision === 'number' ? precision : null, typeof scale === 'number' ? scale : null];
    return field;
  }
  public static decimal(precision?: number, scale?: number) {
    const field = new ModelField<number>();
    field.$type = ['decimal', typeof precision === 'number' ? precision : null, typeof scale === 'number' ? scale : null];
    return field;
  }
  public static boolean() {
    const field = new ModelField<boolean>();
    field.$type = ['boolean'];
    return field;
  }
  public static date() {
    // TODO: You can toggle parsing of dates and get back strings.  How to account for that?
    const field = new ModelField<Date>();
    field.$type = ['date'];
    return field;
  }
  public static datetime(options: { useTz?: boolean; precision?: number } = {}) {
    // TODO: You can toggle parsing of datetimes and get back strings.  How to account for that?
    const field = new ModelField<Date>();
    field.$type = ['datetime', options];
    return field;
  }
  public static time(precision?: number) {
    const field = new ModelField<string>();
    field.$type = ['time', typeof precision === 'number' ? precision : null];
    return field;
  }
  public static timestamp(options: { useTz?: boolean; precision?: number } = {}) {
    const field = new ModelField<Date>();
    field.$type = ['timestamp', options];
    return field;
  }
  public static binary() {
    // TODO: What's the actual types used here?  Have to check all the drivers, I guess.
    const field = new ModelField<Buffer | Uint8Array>();
    field.$type = ['binary'];
    return field;
  }
  public static enum<TValues extends Array<string>>(values: TValues, options: { useNative?: boolean; enumName?: string } = {}) {
    // TODO: Enum...
    const field = new ModelField<TValues>();
    field.$type = ['enum', values, options];
    return field;
  }
  public static json<T = unknown>() {
    // TODO: Should we do anything about json?
    const field = new ModelField<T>();
    field.$type = ['json'];
    return field;
  }
  public static jsonb<T = unknown>() {
    // TODO: Should we do anything about json?
    const field = new ModelField<T>();
    field.$type = ['jsonb'];
    return field;
  }
  public static uuid() {
    const field = new ModelField<string>();
    field.$type = ['uuid'];
    return field;
  }

  // NOTE: name is specified by the field key, not in this class.
  public $wasNamed: string | null = null;
  public $type!: ColumnType;
  public $useInNew: InNew = true as InNew;
  public $unsigned: boolean = false;
  public $comment: string | null = null;
  public $notNullable: NotNullable = false as NotNullable;
  public $default: boolean | number | string | Raw | null = null;
  public $primary: boolean = false;
  public $unique: boolean = false;

  // Fluent mutators ala knex to set the above props.
  // Includes type coercion to maintain exact type params
  // for useful derivation via RecordFieldType<T>.

  public useInNew<TInNew extends boolean>(inNew: TInNew) {
    const next = this as ModelField<TSType, boolean, NotNullable> as ModelField<TSType, TInNew, NotNullable>;
    next.$useInNew = inNew;
    return next;
  }

  public notInNew() {
    return this.useInNew(false);
  }

  public unsigned() {
    this.$unsigned = true;
    return this;
  }

  public comment(commentString: string | null) {
    this.$comment = commentString;
    return this;
  }

  public notNullable() {
    const next = this as ModelField<TSType, InNew, true>;
    next.$notNullable = true;
    return next;
  }

  public default(value: boolean | number | string | Raw) {
    this.$default = value;
    return this;
  }

  public primary() {
    this.$primary = true;
    return this;
  }

  public unique() {
    this.$unique = true;
    return this;
  }

  /**
   * Specifies that this column previously had a different name.
   * This should only be kept around for 1 migration.
   * @param oldName Name of this column in the previous version of the table.
   */
  public wasNamed(oldName: string) {
    this.$wasNamed = oldName;
    return this;
  }
}
