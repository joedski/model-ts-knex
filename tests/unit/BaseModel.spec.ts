import { BaseModel, FooModel } from './fixtures';
import ModelField from '../../lib-ts/ModelField';

describe('BaseModel functionality', () => {
  test('should autogenerate an appropriate table name if none is specified', () => {
    class $RamseysTestKitchenModel extends BaseModel({
      fields: {
        id: ModelField.increments().notInNew(),
        yelling: ModelField.enum(['yes'] as const).notNullable(),
      },
    }) {}

    const RamseysTestKitchenModel = new $RamseysTestKitchenModel();

    expect(RamseysTestKitchenModel.tableName).toBe('ramseys_test_kitchen');
  });

  test('should use the specified name if one is specified', () => {
    class $SomeOtherTestModel extends BaseModel({
      tableName: 'totally_different_name',
      fields: {
        id: ModelField.increments().notInNew(),
      },
    }) {}

    const SomeOtherTestModel = new $SomeOtherTestModel();

    expect(SomeOtherTestModel.tableName).toBe('totally_different_name');
  });

  describe('Query Builder Methods', () => {
    test('should build a basic select with #findWhere', () => {
      expect(FooModel.findWhere({ bar_id: 4 }).toString()).toMatchSnapshot();
      expect(
        FooModel.findWhere({ bar_id: 4 }, ['id']).toString()
      ).toMatchSnapshot();
      expect(
        FooModel.findWhere({ bar_id: 4 }, ['id', 'foo']).toString()
      ).toMatchSnapshot();

      expect(
        FooModel.findWhere(q =>
          q.where({ bar_id: 1 }).orWhere({ bar_id: 2 })
        ).toString()
      ).toMatchSnapshot();
      expect(
        FooModel.findWhere(q => q.where({ bar_id: 1 }).orWhere({ bar_id: 2 }), [
          'id',
        ]).toString()
      ).toMatchSnapshot();
      expect(
        FooModel.findWhere(q => q.where({ bar_id: 1 }).orWhere({ bar_id: 2 }), [
          'id',
          'foo',
        ]).toString()
      ).toMatchSnapshot();
    });

    test('should build a basic select limited to 1 item with #findOneWhere', () => {
      expect(FooModel.findOneWhere({ bar_id: 4 }).toString()).toMatchSnapshot();
      expect(
        FooModel.findOneWhere({ bar_id: 4 }, ['id']).toString()
      ).toMatchSnapshot();
      expect(
        FooModel.findOneWhere({ bar_id: 4 }, ['id', 'foo']).toString()
      ).toMatchSnapshot();

      expect(
        FooModel.findOneWhere(q =>
          q.where({ bar_id: 1 }).orWhere({ bar_id: 2 })
        ).toString()
      ).toMatchSnapshot();
      expect(
        FooModel.findOneWhere(
          q => q.where({ bar_id: 1 }).orWhere({ bar_id: 2 }),
          ['id']
        ).toString()
      ).toMatchSnapshot();
      expect(
        FooModel.findOneWhere(
          q => q.where({ bar_id: 1 }).orWhere({ bar_id: 2 }),
          ['id', 'foo']
        ).toString()
      ).toMatchSnapshot();
    });
  });
});
