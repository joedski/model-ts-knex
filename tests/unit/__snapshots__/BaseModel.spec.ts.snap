// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseModel functionality Query Builder Methods #findWhere should build a basic select with #findWhere 1`] = `"select * from \\"foo\\" where \\"bar_id\\" = 4"`;

exports[`BaseModel functionality Query Builder Methods #findWhere should build a basic select with #findWhere 2`] = `"select \\"id\\" from \\"foo\\" where \\"bar_id\\" = 4"`;

exports[`BaseModel functionality Query Builder Methods #findWhere should build a basic select with #findWhere 3`] = `"select \\"id\\", \\"foo\\" from \\"foo\\" where \\"bar_id\\" = 4"`;

exports[`BaseModel functionality Query Builder Methods #findWhere should build a basic select with #findWhere 4`] = `"select * from \\"foo\\" where (\\"bar_id\\" = 1 or (\\"bar_id\\" = 2))"`;

exports[`BaseModel functionality Query Builder Methods #findWhere should build a basic select with #findWhere 5`] = `"select \\"id\\" from \\"foo\\" where (\\"bar_id\\" = 1 or (\\"bar_id\\" = 2))"`;

exports[`BaseModel functionality Query Builder Methods #findWhere should build a basic select with #findWhere 6`] = `"select \\"id\\", \\"foo\\" from \\"foo\\" where (\\"bar_id\\" = 1 or (\\"bar_id\\" = 2))"`;

exports[`BaseModel functionality Query Builder Methods should build a basic insert with #insertOne 1`] = `"insert into \\"foo\\" (\\"bar_id\\", \\"foo\\") values (2, 25)"`;

exports[`BaseModel functionality Query Builder Methods should build a basic insert with #insertOne 2`] = `"insert into \\"foo\\" (\\"bar_id\\", \\"foo\\") values (2, 25) returning \\"id\\""`;

exports[`BaseModel functionality Query Builder Methods should build a basic select limited to 1 item with #findOneWhere 1`] = `"select * from \\"foo\\" where \\"bar_id\\" = 4 limit 1"`;

exports[`BaseModel functionality Query Builder Methods should build a basic select limited to 1 item with #findOneWhere 2`] = `"select \\"id\\" from \\"foo\\" where \\"bar_id\\" = 4 limit 1"`;

exports[`BaseModel functionality Query Builder Methods should build a basic select limited to 1 item with #findOneWhere 3`] = `"select \\"id\\", \\"foo\\" from \\"foo\\" where \\"bar_id\\" = 4 limit 1"`;

exports[`BaseModel functionality Query Builder Methods should build a basic select limited to 1 item with #findOneWhere 4`] = `"select * from \\"foo\\" where (\\"bar_id\\" = 1 or (\\"bar_id\\" = 2)) limit 1"`;

exports[`BaseModel functionality Query Builder Methods should build a basic select limited to 1 item with #findOneWhere 5`] = `"select \\"id\\" from \\"foo\\" where (\\"bar_id\\" = 1 or (\\"bar_id\\" = 2)) limit 1"`;

exports[`BaseModel functionality Query Builder Methods should build a basic select limited to 1 item with #findOneWhere 6`] = `"select \\"id\\", \\"foo\\" from \\"foo\\" where (\\"bar_id\\" = 1 or (\\"bar_id\\" = 2)) limit 1"`;
